libraries/chain/fork_database.cpp

```diff
@@ -533,6 +440,11 @@ namespace eosio::chain {
 
    template<class BSP>
    BSP fork_database_impl<BSP>::search_on_head_branch_impl( uint32_t block_num, include_root_t include_root ) const {
+      auto head = head_impl(include_root_t::no);
+      if (!head && include_root == include_root_t::yes && block_num == root->block_num())
+         return root;
+      if (!head)
+         return {};
       return search_on_branch_impl(head->id(), block_num, include_root);
```

Comment:
I think this is equivalent and somewhat simpler.

```suggestion
   auto head = head_impl(include_root);
   if (!head || head == root)
      return head;
   return search_on_branch_impl(head->id(), block_num, include_root);
```